Generate changes for staged repos (apimachinery, apiserver)

smarterclayton opened the pull request:
Restores OpenAPI types, deepcopy, and bazel for meta/v1 types.

Fixes #39912

smarterclayton committed:
Generate for staging directories that are not client-go

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39916)
<!-- Reviewable:end -->

smarterclayton commented:
@mikedanese need some gazel advice.  apimachinery and apiserver are something we haven't tried before (with predictable results) - staging/src is the authoritative source, they are symlinked into vendor, and will be eventually copied out.  However, unlikely client-go they aren't an artifact of the main tree, they are the canonical source.  Since they're symlinked into the tree, the BUILD files show up under vendor which seems to make bazel angry.

smarterclayton commented:
Gazel keeps trying to insert a link to staging (which is not part of the source tree) - probably because apimachinery is symlinked into vendor and *maybe* gazel is chasing the symlink out of vendor?

```
     deps = [
         "//pkg/api/resource:go_default_library",
         "//pkg/util/intstr:go_default_library",
+        "//staging/src/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",
         "//vendor:github.com/go-openapi/spec",
-        "//vendor:k8s.io/apimachinery/pkg/apis/meta/v1",
         "//vendor:k8s.io/apimachinery/pkg/genericapiserver/openapi/common",
     ],
```

mikedanese commented:
@smarterclayton it looks like this is caused by your import:
```
   28         v1 "k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1"
   29 )
```
in pkg/generated/openapi/zz_generated.openapi.go. Should that be just k8s.io/apimachinery/pkg/apis/meta/v1?

smarterclayton commented:
It should be - thanks, didn't spot that in the maze of twisty packages.  Looks like the easy fix for the staging/src dependencies is not going to be easy.

smarterclayton committed:
Strip k8s gen tags from client-go

Prevents client-go from being part of generation now that vendored
dependencies are considered in the chain.

smarterclayton committed:
generated: bump staging without codegen flags

smarterclayton committed:
Mark meta/v1 as writable temporarily

Allow openapi-gen tag to be removed from the directory.

smarterclayton commented:
@sttts @deads2k the approach here is to make certain vendor directories candidates for standard generation, and to also relax constraints on gengo that prevent it from processing nested packages.  I think this is the minimal change that reenables generation of meta/v1 for the group types, although I can imagine other paths (like doing generation within staging/src as a sub gopath first).  Even in that model, some packages in vendor will have to be read (openapi, for instance, needs to include types it finds in nested packages).  Staging was updated to strip generation tags because client-go is now a candidate (being in the vendor directory).  In general we would not generate from client-go today, but in the future we might.

The linked gengo PR must be merged first, then I'll update to bump.

smarterclayton committed:
generated: swagger

sttts commented on staging/copy.sh:
```diff
@@ -130,6 +131,13 @@ sed -i "s/kubernetes_build_info/kubernetes_build_info_copy/g" "${CLIENT_REPO_TEM
 find "${CLIENT_REPO_TEMP}" -type f -name "generated.pb.go" -print0 | xargs -0 sed -i "s/k8s\.io\.kubernetes/k8s.io.client-go/g"
 
+# strip all generator tags from client-go
+find "${CLIENT_REPO_TEMP}" -type f -name "*.go" -print0 | xargs -0 sed -i "/^\/\/ +k8s:openapi-gen=true/d"
```
`sed -i "/^\/\/ +k8s:[a-z]*-gen=true/d"`

sttts replied:
The + is interpreted I guess and must be escaped.

smarterclayton replied:
I don't want to pick all of them, because some gen's might actually need to stay.

sttts commented on staging/copy.sh:
```diff
@@ -42,6 +42,7 @@ CLIENT_REPO="${MAIN_REPO}/staging/src/${CLIENT_REPO_FROM_SRC}"
 CLIENT_REPO_TEMP="${MAIN_REPO}/staging/src/${CLIENT_REPO_TEMP_FROM_SRC}"
 
 DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+sedi="-i"
```
is it used anywhere?

Looks like `sed -i=""` works for both gnu and OSX.

smarterclayton replied:
Ok

sttts commented on hack/make-rules/helpers/cache_go_dirs.sh:
```diff
@@ -36,7 +36,8 @@ trap "rm -f '${CACHE}'" HUP INT TERM ERR
 # Example:
 #   kfind -type f -name foobar.go
 function kfind() {
-    find .                         \
```
Please add a comment why this is the correct path list. Looks like only Makefile.generated_files calls cache_go_dir.sh, luckily.

deads2k commented:
There's a "do not merge" commit in this pull.  How close is this?

deads2k commented on pkg/generated/openapi/zz_generated.openapi.go:
```diff
@@ -8280,24 +8281,7 @@ var OpenAPIDefinitions *common.OpenAPIDefinitions = &common.OpenAPIDefinitions{
-	"v1.Time": {
-		Schema: spec.Schema{
-			SchemaProps: spec.SchemaProps{
-				Description: "Time is a wrapper around time.Time which supports correct marshaling to YAML and JSON.  Wrappers are provided for many of the factory methods that the time package offers.",
```
looks like something went wrong.

smarterclayton replied:
No, this is it going right.  There is an override that was being missed.

deads2k commented:
Found pkg/runtime/.readonly, but files changed compared to "master" branch.

smarterclayton commented:
And jenkins breaks

deads2k commented:
@k8s-bot test this

smarterclayton committed:
bump(k8s.io/gengo):3c6a809462caf39389d70d9ea787ed24c5acffed

deads2k commented:
conflicts

smarterclayton committed:
generated: OpenAPI is now correct

deads2k commented:
lgtm
